activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/DefaultJobMessageProducer.java

```
@@ -43,31 +42,57 @@ public DefaultJobMessageProducer(BinderAwareChannelResolver resolver,
     }
 
     @Override
-    public void sendMessage(String destination, Job job) {
-        eventPublisher.publishEvent(new JobMessageEvent(job.getId(), destination, job));
+    public void sendMessage(@NonNull String destination, @NonNull Job job) {
+        Assert.isTrue(TransactionSynchronizationManager.isSynchronizationActive(), "requires active transaction synchronization");
+        Assert.hasLength(job.getId(), "job id must not be empty");
+        Assert.hasLength(destination, "destination must not be empty");
+        
+        // Let's try to resolve message channel while inside main Activiti transaction to minimize infrastructure errors 
+        MessageChannel messageChannel = resolver.resolveDestination(destination);
+
+        // Let's send message right after the main transaction has successfully committed. 
+        TransactionSynchronizationManager.registerSynchronization(new JobMessageTransactionSynchronization(destination, 
+                                                                                                           messageChannel, 
+                                                                                                           job));
+    }
+    
+    protected Message<String> buildMessage(Job job) {
+        return MessageBuilder.withPayload(job.getId())
+                // TODO set headers?  
+                .build();
     }
     
-    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
-    public void on(@NonNull JobMessageEvent event) {
-        logger.debug("On JobMessageEvent: {}", event);
+    class JobMessageTransactionSynchronization implements TransactionSynchronization {
```

Is there any specific reason to use `TransactionSynchronization` instead of `@TransactionalEventListener`? I thought `TransactionSynchronization` was the old way of doing it and `@TransactionalEventListener` the new way. In [this blog post](http://pillopl.github.io/reliable-domain-events/), for instance, they are mentioning that code could be simplified by using `@TransactionalEventListener`.

---

I favor a more direct approach to transaction synchronization to be able to fail fast inside main Activiti transaction by validating job attributes and resolving destination channel and then register `TransactionSynchronization` handler to send message after successful commit. The `TransactionalEventListener` uses `TransactionSynchronization` to do the same.